File: cwclock-mobile/src/screens/clients/ClientFormScreen.js
import React, { useEffect, useMemo, useState } from "react";
import { Alert, StyleSheet } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import FormField from "../../components/FormField";
import SelectField from "../../components/SelectField";
import ToggleRow from "../../components/ToggleRow";
import Button from "../../components/Button";
import ErrorBanner from "../../components/ErrorBanner";
import { useI18n } from "../../i18n/I18nContext";
import { apiErrorMessage } from "../../i18n/translate";
import theme from "../../theme";
import { listCountriesApi } from "../../redux/countries/countries.actions";
import { createClientApi, updateClientApi, deleteClientApi } from "../../redux/clients/clients.actions";

const emptyFields = {
  name: "",
  email: "",
  contactName: "",
  address: "",
  postalCode: "",
  city: "",
  country: "",
  vatNumber: "",
  vatRate: "",
  identificationNumber: "",
  purchaseOrder: "",
  hoursPerDay: "",
  dailyRate: "",
  sendReportsWithInvoice: false,
};

// ClientFormScreen: create (route.params.client absent) or edit (present).
// Ported from cwclock-ui's EditClientModal.jsx, trimmed to the core fields -
// no country-specific identification fields (SIREN/SIRET/NAF/MF, see
// useCountryFields.js - a generic "Identification number" field covers that
// need here), invoiceEmails/vatDischargeMotive, or cross-organization
// transfer (ai-instruct-106 scope).
const ClientFormScreen = ({ route, navigation }) => {
  const { t, locale } = useI18n();
  const dispatch = useDispatch();
  const { orgId } = useSelector((state) => state.session);
  const { items: countries } = useSelector((state) => state.countries);
  const editingClient = route.params?.client;

  const [fields, setFields] = useState(() =>
    editingClient
      ? {
          name: editingClient.name || "",
          email: editingClient.email || "",
          contactName: editingClient.contactName || "",
          address: editingClient.address || "",
          postalCode: editingClient.postalCode || "",
          city: editingClient.city || "",
          country: editingClient.country || "",
          vatNumber: editingClient.vatNumber || "",
          vatRate: editingClient.vatRate ? String(editingClient.vatRate) : "",
          identificationNumber: editingClient.identificationNumber || "",
          purchaseOrder: editingClient.purchaseOrder || "",
          hoursPerDay: editingClient.hoursPerDay ? String(editingClient.hoursPerDay) : "",
          dailyRate: editingClient.dailyRate != null ? String(editingClient.dailyRate) : "",
          sendReportsWithInvoice: editingClient.sendReportsWithInvoice || false,
        }
      : emptyFields
  );
  const [error, setError] = useState(null);
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    dispatch(listCountriesApi());
  }, [dispatch]);

  const countryItems = useMemo(() => countries.map((c) => ({ value: c.iso, label: c.name })), [countries]);

  const setField = (key, value) => setFields((f) => ({ ...f, [key]: value }));

  const toNumber = (v) => (v === "" ? undefined : Number(v));

  const handleSave = async () => {
    setError(null);
    setSaving(true);
    const payload = {
      ...fields,
      vatRate: toNumber(fields.vatRate),
      hoursPerDay: toNumber(fields.hoursPerDay),
      dailyRate: toNumber(fields.dailyRate),
    };
    try {
      if (editingClient) {
        await dispatch(updateClientApi(orgId, editingClient.id, payload));
      } else {
        await dispatch(createClientApi(orgId, payload));
      }
      navigation.goBack();
    } catch (e) {
      setError(apiErrorMessage(e, locale));
    } finally {
      setSaving(false);
    }
  };

  const handleDelete = () => {
    Alert.alert(t("clients.deleteClientTitle"), t("clients.deleteClientBody", { name: editingClient.name }), [
      { text: t("common.cancel"), style: "cancel" },
      {
        text: t("common.delete"),
        style: "destructive",
        onPress: async () => {
          await dispatch(deleteClientApi(orgId, editingClient.id));
          navigation.goBack();
        },
      },
    ]);
  };

  return (
    <Screen>
      <FormField label={t("clients.name")} value={fields.name} onChangeText={(v) => setField("name", v)} />
      <FormField
        label={t("clients.email")}
        value={fields.email}
        onChangeText={(v) => setField("email", v)}
        autoCapitalize="none"
        keyboardType="email-address"
      />
      <FormField
        label={t("clients.contactName")}
        value={fields.contactName}
        onChangeText={(v) => setField("contactName", v)}
      />
      <FormField label={t("clients.address")} value={fields.address} onChangeText={(v) => setField("address", v)} />
      <FormField
        label={t("clients.postalCode")}
        value={fields.postalCode}
        onChangeText={(v) => setField("postalCode", v)}
      />
      <FormField label={t("clients.city")} value={fields.city} onChangeText={(v) => setField("city", v)} />
      <SelectField
        label={t("clients.country")}
        value={fields.country}
        onValueChange={(v) => setField("country", v)}
        items={countryItems}
        placeholder={t("clients.country")}
      />
      <FormField
        label={t("clients.vatNumber")}
        value={fields.vatNumber}
        onChangeText={(v) => setField("vatNumber", v)}
      />
      <FormField
        label={t("clients.vatRate")}
        value={fields.vatRate}
        onChangeText={(v) => setField("vatRate", v)}
        keyboardType="decimal-pad"
      />
      <FormField
        label={t("clients.identificationNumber")}
        value={fields.identificationNumber}
        onChangeText={(v) => setField("identificationNumber", v)}
      />
      <FormField
        label={t("clients.purchaseOrder")}
        value={fields.purchaseOrder}
        onChangeText={(v) => setField("purchaseOrder", v)}
      />
      <FormField
        label={t("clients.hoursPerDay")}
        value={fields.hoursPerDay}
        onChangeText={(v) => setField("hoursPerDay", v)}
        keyboardType="decimal-pad"
      />
      <FormField
        label={t("clients.dailyRate")}
        value={fields.dailyRate}
        onChangeText={(v) => setField("dailyRate", v)}
        keyboardType="decimal-pad"
      />
      <ToggleRow
        label={t("clients.sendReportsWithInvoice")}
        value={fields.sendReportsWithInvoice}
        onValueChange={(v) => setField("sendReportsWithInvoice", v)}
      />
      <ErrorBanner message={error} />
      <Button
        title={t("common.save")}
        onPress={handleSave}
        loading={saving}
        disabled={!fields.name.trim() || !fields.country}
        style={editingClient ? styles.saveButton : undefined}
      />
      {editingClient && <Button title={t("common.delete")} variant="danger" onPress={handleDelete} />}
    </Screen>
  );
};

const styles = StyleSheet.create({
  saveButton: {
    marginBottom: theme.spacing(1.5),
  },
});

export default ClientFormScreen;

File: cwclock-mobile/lib/screens/clients/client_form_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../api/api_client.dart';
import '../../i18n/app_localizations.dart';
import '../../models/client.dart';
import '../../providers/clients_provider.dart';
import '../../providers/countries_provider.dart';
import '../../providers/locale_provider.dart';
import '../../providers/session_provider.dart';
import '../../theme.dart';
import '../../widgets/app_button.dart';
import '../../widgets/app_screen.dart';
import '../../widgets/error_banner.dart';
import '../../widgets/form_field.dart';
import '../../widgets/select_field.dart';
import '../../widgets/toggle_row.dart';
import '../../widgets/app_top_bar.dart';

/// Create (client is null) or edit (present). Ported from
/// src/screens/clients/ClientFormScreen.js.
class ClientFormScreen extends ConsumerStatefulWidget {
  final Client? client;

  const ClientFormScreen({super.key, this.client});

  @override
  ConsumerState<ClientFormScreen> createState() => _ClientFormScreenState();
}

class _ClientFormScreenState extends ConsumerState<ClientFormScreen> {
  late ClientFields _fields =
      widget.client != null ? ClientFields.fromClient(widget.client!) : const ClientFields();
  String? _error;
  bool _saving = false;

  @override
  void initState() {
    super.initState();
    Future.microtask(() => ref.read(countriesProvider.notifier).listCountries());
  }

  Future<void> _handleSave() async {
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;
    final locale = ref.read(localeProvider);

    setState(() {
      _error = null;
      _saving = true;
    });
    try {
      if (widget.client != null) {
        await ref.read(clientsProvider.notifier).updateClient(orgId, widget.client!.id, _fields.toJson());
      } else {
        await ref.read(clientsProvider.notifier).createClient(orgId, _fields.toJson());
      }
      if (mounted) Navigator.of(context).pop();
    } catch (e) {
      setState(() => _error = apiErrorMessage(asApiException(e), locale));
    } finally {
      if (mounted) setState(() => _saving = false);
    }
  }

  void _handleDelete() {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null || widget.client == null) return;

    showDialog(
      context: context,
      builder: (dialogContext) => AlertDialog(
        title: Text(t('clients.deleteClientTitle')),
        content: Text(t('clients.deleteClientBody', {'name': widget.client!.name})),
        actions: [
          TextButton(onPressed: () => Navigator.pop(dialogContext), child: Text(t('common.cancel'))),
          TextButton(
            onPressed: () async {
              Navigator.pop(dialogContext);
              await ref.read(clientsProvider.notifier).deleteClient(orgId, widget.client!.id);
              if (!mounted) return;
              Navigator.of(context).pop();
            },
            child: Text(t('common.delete'), style: TextStyle(color: AppColors.of(context).danger)),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final countries = ref.watch(countriesProvider);
    final countryItems = [for (final c in countries) SelectItem(c.iso, c.name)];
    final f = _fields;

    return Scaffold(
      appBar: AppTopBar(title: widget.client != null ? t('clients.editClient') : t('clients.addClient')),
      body: AppScreen(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            AppFormField(label: t('clients.name'), value: f.name, onChanged: (v) => setState(() => _fields = f.copyWith(name: v))),
            AppFormField(
              label: t('clients.email'),
              value: f.email,
              onChanged: (v) => setState(() => _fields = f.copyWith(email: v)),
              keyboardType: TextInputType.emailAddress,
              textCapitalization: TextCapitalization.none,
            ),
            AppFormField(
              label: t('clients.contactName'),
              value: f.contactName,
              onChanged: (v) => setState(() => _fields = f.copyWith(contactName: v)),
            ),
            AppFormField(
              label: t('clients.address'),
              value: f.address,
              onChanged: (v) => setState(() => _fields = f.copyWith(address: v)),
            ),
            AppFormField(
              label: t('clients.postalCode'),
              value: f.postalCode,
              onChanged: (v) => setState(() => _fields = f.copyWith(postalCode: v)),
            ),
            AppFormField(
              label: t('clients.city'),
              value: f.city,
              onChanged: (v) => setState(() => _fields = f.copyWith(city: v)),
            ),
            AppSelectField(
              label: t('clients.country'),
              value: f.country,
              onChanged: (v) => setState(() => _fields = f.copyWith(country: v)),
              items: countryItems,
              placeholder: t('clients.country'),
            ),
            AppFormField(
              label: t('clients.vatNumber'),
              value: f.vatNumber,
              onChanged: (v) => setState(() => _fields = f.copyWith(vatNumber: v)),
            ),
            AppFormField(
              label: t('clients.vatRate'),
              value: f.vatRate,
              onChanged: (v) => setState(() => _fields = f.copyWith(vatRate: v)),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
            ),
            AppFormField(
              label: t('clients.identificationNumber'),
              value: f.identificationNumber,
              onChanged: (v) => setState(() => _fields = f.copyWith(identificationNumber: v)),
            ),
            AppFormField(
              label: t('clients.purchaseOrder'),
              value: f.purchaseOrder,
              onChanged: (v) => setState(() => _fields = f.copyWith(purchaseOrder: v)),
            ),
            AppFormField(
              label: t('clients.hoursPerDay'),
              value: f.hoursPerDay,
              onChanged: (v) => setState(() => _fields = f.copyWith(hoursPerDay: v)),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
            ),
            AppFormField(
              label: t('clients.dailyRate'),
              value: f.dailyRate,
              onChanged: (v) => setState(() => _fields = f.copyWith(dailyRate: v)),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
            ),
            AppToggleRow(
              label: t('clients.sendReportsWithInvoice'),
              value: f.sendReportsWithInvoice,
              onChanged: (v) => setState(() => _fields = f.copyWith(sendReportsWithInvoice: v)),
            ),
            ErrorBanner(message: _error),
            AppButton(
              title: t('common.save'),
              onPressed: f.name.trim().isEmpty || f.country.isEmpty ? null : _handleSave,
              loading: _saving,
              margin: EdgeInsets.only(bottom: AppSpacing.of(1.5)),
            ),
            if (widget.client != null)
              AppButton(title: t('common.delete'), variant: AppButtonVariant.danger, onPressed: _handleDelete),
          ],
        ),
      ),
    );
  }
}
